pkg/errors/crosschain_error_message.go

```diff
@@ -0,0 +1,185 @@
+
+// Zevm specific fields. These fields are only available if the outbound or revert was involved in a ZEVM transaction, as a deposit
+// Contract: Contract called
+// Args: Arguments provided to the call
+// Method: Contract method
+// RevertReason: Reason for the revert if available
+type CCTXErrorMessage struct {
+	Message      string `json:"message"`
+	Error        string `json:"error"`
+	Method       string `json:"method"`
+	Contract     string `json:"contract"`
+	Args         string `json:"args"`
+	RevertReason string `json:"revert_reason"`
+}
+
+// NewZEVMErrorMessage creates a new CCTXErrorMessage,
+// which contains additional fields which are specific to ZEVM calls only
+func NewZEVMErrorMessage(
+	method string,
+	contract common.Address,
+	args interface{},
+	message string,
+	err error,
+) CCTXErrorMessage {
+	c := CCTXErrorMessage{
+		Method:   method,
+		Contract: contract.String(),
+		Args:     fmt.Sprintf("%v", args),
+		Message:  message,
+	}
+	if err != nil {
+		c.Error = err.Error()
+	}
+	return c
+}
+
+// NewCCTXErrorJSONMessage creates a new CCTXErrorMessage struct and returns it as a JSON string.
+// The error field can be of the following types
+// 1 - Nil : If the error is nil, we don't need to do anything
+// 2 - ErrorString : If the error is a string, we add the error and Pack it into a CCTXErrorMessage struct
+// 3 - CCTXErrorMessage : If the error is already a CCTXErrorMessage, we should unpack it and return the JSON string,
+// 4 - If it the error is a chain of errors, we should unpack each error and add it to the CCTXErrorMessage struct error field
+// 5 - If the error is a chain of errors and one of the errors is a CCTXErrorMessage, we should unpack the CCTXErrorMessage and return the JSON string and add the errors into the error field
+// This function does not return an error, if the marshalling fails, it returns a string representation of the CCTXErrorMessage
+func NewCCTXErrorJSONMessage(message string, newError error) string {
+	m := &CCTXErrorMessage{
+		Message: message,
+	}
+
+	if newError != nil {
+		errorLists := SplitErrorMessage(newError.Error())
+
+		for _, e := range errorLists {
+			parsed, err := ParseCCTXErrorMessage(e)
+			switch {
+			// parsing failed so this is not a CCTXErrorMessage json string
+			case err != nil:
+				{
+					m.WrapError(e)
+				}
+			// parsing succeeded, this is a CCTXErrorMessage unpack it
+			default:
+				{
+					m.Message = parsed.Message
+					m.Method = parsed.Method
+					m.Contract = parsed.Contract
+					m.Args = parsed.Args
+					m.RevertReason = parsed.RevertReason
+					m.WrapError(parsed.Error)
+				}
+			}
+		}
+	}
+
+	jsonString, err := m.ToJSON()
+	if err != nil {
+		return fmt.Sprintf("json marshalling failed %s,%s", err.Error(), m.String())
+	}
+	return jsonString
+}
+
+// SplitErrorMessage splits the error message into parts it treats a json formatted error message as a single part
+func SplitErrorMessage(input string) []string {
```

Comment:
THe current implementation only hanldes 1 json object in the error chain , it cna be easily modified to handle multiple json objects `CCTXErrorMessage` . 

However that would add complexity to this function and also to `NewCCTXErrorJSONMessage`. 

If it's needed, I can refactor this function